src/settings/settings_manager.cpp

```diff
@@ -0,0 +1,175 @@
+  std::string old_value = GetString(param);
+  SetValue(param, type::ValueFactory::GetVarcharValue(value.c_str()));
+  callback_fn callback = callback_map_.find(param)->second;
+  std::string new_value(value);
+  callback(static_cast<void *>(&old_value), static_cast<void *>(&new_value));
+}
+
+const std::string SettingsManager::GetInfo() {
+  /*
+  const uint32_t box_width = 72;
+  const std::string title = "PELOTON SETTINGS";
+
+  std::string info;
+  info.append(StringUtil::Format("%*s\n", box_width / 2 + title.length() / 2,
+                                 title.c_str()));
+  info.append(StringUtil::Repeat("=", box_width)).append("\n");
+
+  // clang-format off
+  info.append(StringUtil::Format("%34s:   %-34i\n", "Port", GetInt(Param::port)));
+  info.append(StringUtil::Format("%34s:   %-34s\n", "Socket Family", GetString(Param::socket_family).c_str()));
+  info.append(StringUtil::Format("%34s:   %-34s\n", "Statistics", GetInt(Param::stats_mode) ? "enabled" : "disabled"));
+  info.append(StringUtil::Format("%34s:   %-34i\n", "Max Connections", GetInt(Param::max_connections)));
+  info.append(StringUtil::Format("%34s:   %-34s\n", "Index Tuner", GetBool(Param::index_tuner) ? "enabled" :
+  "disabled")); info.append(StringUtil::Format("%34s:   %-34s\n", "Layout Tuner", GetBool(Param::layout_tuner) ?
+  "enabled" : "disabled")); info.append(StringUtil::Format("%34s:   (queue size %i, %i threads)\n", "Worker Pool",
+  GetInt(Param::monoqueue_task_queue_size), GetInt(Param::monoqueue_worker_pool_size)));
+  info.append(StringUtil::Format("%34s:   %-34s\n", "Parallel Query Execution", GetBool(Param::parallel_execution) ?
+  "enabled" : "disabled")); info.append(StringUtil::Format("%34s:   %-34i\n", "Min. Parallel Table Scan Size",
+  GetInt(Param::min_parallel_table_scan_size))); info.append(StringUtil::Format("%34s:   %-34s\n", "Code-generation",
+  GetBool(Param::codegen) ? "enabled" : "disabled")); info.append(StringUtil::Format("%34s:   %-34s\n", "Print IR
+  Statistics", GetBool(Param::print_ir_stats) ? "enabled" : "disabled")); info.append(StringUtil::Format("%34s:
+  %-34s\n", "Dump IR", GetBool(Param::dump_ir) ? "enabled" : "disabled")); info.append(StringUtil::Format("%34s:
+  %-34i\n", "Optimization Timeout", GetInt(Param::task_execution_timeout))); info.append(StringUtil::Format("%34s:
+  %-34i\n", "Number of GC threads", GetInt(Param::gc_num_threads)));
+  // clang-format on
+
+  return StringBoxUtil::Box(info);
+   */
+  return "";
+}
+
+void SettingsManager::ShowInfo() { /*LOG_INFO("\n%s\n", GetInfo().c_str());*/
+}
+
+type::Value SettingsManager::GetValue(Param param) {
+  auto param_info = param_map_.find(param);
+  return param_info->second.value;
+}
+
+void SettingsManager::SetValue(Param param, const type::Value &value) {
+  ParamInfo& param_info = param_map_.find(param)->second;
+
+  if (!param_info.is_mutable) throw SETTINGS_EXCEPTION((param_info.name + " is not mutable.").c_str());
```

Comment:
Is it the correct behavior to throw an exception? Is the upper level users responsible for catching these exceptions?

Reply:
Maybe you are right. It would be better to LOG_ERROR and tell the caller that this `set` failed.